### DISCUSSION TITLE
What should I use: tags or custom branches

### DISCUSSION THREAD

**Discussion by malcolm061990** [prometheus-operator/kube-prometheus#1411] [Q&A] [answered]
Hi guys.
First of all big thanks for this project. It's amazing :)
But for now I can't understand what source code can and should I use.
There are git tags:
v0.6.0
v0.7.0
v0.8.0

There are custom branches:
release-0.7
release-0.8
release-0.9

As a git user I consider that git tag is a version of ready to production product and branch is just a branch.
Using this default logic I use only git tags. The latest tag is v0.8.0 from April 27 of 2021 but as I see the compatibility [matrix](https://github.com/prometheus-operator/kube-prometheus#kubernetes-compatibility-matrix) talks about branches, not about tags.
So what should I use: tags or custom branches?

**Comment by paulfantom** [comment, marked as answer]
in short: branches ;)

there are bugfixes landing on branches, but we don't release tags for those.

**Reply by malcolm061990** [reply]
> in short: branches ;)
> 
> there are bugfixes landing on branches, but we don't release tags for those.

Thanks for your reply. But I don't understand why this mechanism was chosen.
On general the best practice is to use tags for ready-to-production code. Tags are immutable, branches are not.
There are many evidences for that.

**Reply by paulfantom** [reply]
In the beginning this project was meant to be used as a library with a rolling release and use only `master/main` branch. To accomplish this we wrote and started using [jsonnet-bundler](https://github.com/jsonnet-bundler/jsonnet-bundler) which in turn is using a locking mechanism based on git commits, so locking it twice (version and commit msg) didn't make any sense at that point in time. Later, during development, branches were introduced to allow downstream users having long term support for older kubernetes and OpenShift versions.

That said we are open to discuss improvements in this area. Including possibility to change how project is versioned and with what release cadence. However bear in mind that the project is not maintained by a huge number of people and we need a solution to manage [5 separate branches](https://github.com/prometheus-operator/kube-prometheus/blob/main/.github/workflows/versions.yaml#L13-L17) at the same time.

>  Tags are immutable, branches are not. 

Commits are also immutable and this is what we are using if immutability is required. This way you are 100% sure your reference wasn't deleted and pushed again pointing to other place (as one could do with tags).
